src/Polysemy/History.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TemplateHaskell     #-}
+
+module Polysemy.History
+  ( -- * Effect
+    History (..)
+
+    -- * Actions
+  , undo
+  , redo
+  , putAndForget
+
+    -- * Interpretations
+  , runHistory
+  ) where
+
+import Polysemy
+import Polysemy.State
+import Polysemy.Tagged
+
+
+data History s m a where
+  Undo :: History s m ()
+  Redo :: History s m ()
+  PutAndForget :: s -> History s m ()
+
+makeSem ''History
+
+
+runHistory
+    :: forall s r a
+     . Member (State s) r
+    => Sem (History s ': r) a
+    -> Sem r a
+runHistory m = do
+  s <- get
+  evalState (Zipper [] s [])
```

Comment:
If an error effect comes between `History` and `State s`, then it's possible to throw exceptions that will revert the record-keeping of `History`, but not affect the targeted `State s`, sabotaging consistency.

Reply:
haddock: "Make sure to run this immediately before running your state!!!"?

Reply:
If we want that restriction, we could do:
```haskell
runHistory :: s -> Sem (History s ': State s ': r) a -> Sem r (s, a)
runHistory s = runState s . ...
```

Reply:
The plus side to that is that `historyToIO` would lose its problems too.
```haskell
historyToIO :: s -> Sem (History s ': State s ': r) a -> Sem r (s, a)
historyToIO s = stateToIO s . ...
```

Reply:
but then you can't mix and match your state :(

Reply:
Oh yeah. Damn.

Reply:
> haddock: "Make sure to run this immediately before running your state!!!"?
How is this different than forcing `History` to come before `State` in the stack?

Maybe I didn't understand what you mean by "Mix and match the state"

Reply:
One example is that it's possible to interpret a `State s` in terms of `State t` if you have a `Lens' t s` (although it's not actually offered in the library):

```haskell
mapState :: Member (State t) r => Lens' t s -> Sem (State s ': r) a -> Sem r a
mapState lens = interpret $ \case
  Put s -> do
    t <- get
    put $ set lens s t
  Get -> view lens <$> get
```
Hence, you can mix and match smaller pieces of state with larger pieces of state. These kinds of transformations are impossible if we force one part of the state to be interpreted immediately as part of interpreting `History`. 
Admittedly, perhaps @isovector meant something different.

Reply:
more generally, i meant "there are lots of state interpretations, why should we privilege `runState` and `stateToIO`"?

Reply:
Oh okay, so basically there are other useful cases where you might want to add an interpretation between `History` and `State`, right?

Reply:
my design principle is "people in the future are going to be smarter than i am, so don't make arbitrary restrictions in what they can do just because i don't see it"

Reply:
I see, that's actually a pretty good design principle!